### PR TITLE
Support javax.net.ssl.[Key|Trust]Manager in Key and Trust credentials.

### PR DESCRIPTION
**Pull request by spockz**
Support `javax.net.ssl.[Key|Trust]Manager` in Key and Trust credentials.

Problem

Proper H2 support is only provided when configuring the Http client with the new `Ssl[Client|Server]Configuration` which takes `KeyCredentials` and `TrustCredentials`. Before this PR those could only be constructed from PEM files. We use .jks files to ship our credentials which are not usable with the `KeyCredentials` and `TrustCredentials`.

For more background see #650.

Solution

I've added extra case classes for both `KeyCredentials` and `TrustCredentials` and made sure they are passed through to netty.

Please feel free to (nit)pick on the names for the case classes. :)

### CODE REVIEW

File: finagle-core/src/main/java/com/twitter/finagle/ssl/KeyCredentialsConfig.java [hunk 1]
```diff
@@ -33,4 +34,11 @@ public static KeyCredentials certKeyAndChain(
     return new KeyCredentials.CertKeyAndChain(certificateFile, keyFile, caCertificateFile);
   }
 
+  /**
+   * See {@link KeyCredentials.FromKeyManager}
+   * @see KeyCredentials.FromKeyManager // Refactor to use @see instead?
+   */
+  public static KeyCredentials fromKeyManagerFactory(KeyManagerFactory keyManagerFactory) {
```
Comment (kevinoliver) [inline finagle-core/src/main/java/com/twitter/finagle/ssl/KeyCredentialsConfig.java]: i think the naming pattern in this file would make this method be called "keyManagerFactory`.

File: finagle-core/src/main/java/com/twitter/finagle/ssl/TrustCredentialsConfig.java [hunk 2]
```diff
@@ -30,4 +31,11 @@ public static TrustCredentials certCollection(File file) {
     return new TrustCredentials.CertCollection(file);
   }
 
+  /**
+   * See {@link TrustCredentials.FromTrustManagerFactory}
+   */
+  public static TrustCredentials fromTrustManagerFactory(TrustManagerFactory trustManagerFactory) {
```
Comment (kevinoliver) [inline finagle-core/src/main/java/com/twitter/finagle/ssl/TrustCredentialsConfig.java]: perhaps a similar comment here on naming, but i defer to @ryanoneill

File: finagle-core/src/main/scala/com/twitter/finagle/ssl/KeyCredentials.scala [hunk 3]
```diff
@@ -48,4 +50,10 @@ object KeyCredentials {
    */
   case class CertKeyAndChain(certificateFile: File, keyFile: File, caCertificateFile: File)
       extends KeyCredentials
+
+  /**
+   *
```
Comment (kevinoliver) [inline finagle-core/src/main/scala/com/twitter/finagle/ssl/KeyCredentials.scala]: lets fill this comment in.

File: finagle-core/src/main/scala/com/twitter/finagle/ssl/KeyCredentials.scala [hunk 4]
```diff
@@ -48,4 +50,10 @@ object KeyCredentials {
...
+   * @param keyManagerFactory
+   */
+  case class FromKeyManager(keyManagerFactory: KeyManagerFactory) extends KeyCredentials
```
Comment (kevinoliver) [inline finagle-core/src/main/scala/com/twitter/finagle/ssl/KeyCredentials.scala]: same sentiment here on the name

File: finagle-core/src/main/scala/com/twitter/finagle/ssl/SslConfigurations.scala [hunk 5]
```diff
@@ -21,18 +21,20 @@ private[ssl] object SslConfigurations {
   def getKeyManagers(keyCredentials: KeyCredentials): Option[Array[KeyManager]] =
     keyCredentials match {
       case KeyCredentials.Unspecified => None
-      case KeyCredentials.CertAndKey(certFile, keyFile) =>
+      case KeyCredentials.CertAndKey(certFile, keyFile)     =>
```
Comment (kevinoliver) [inline finagle-core/src/main/scala/com/twitter/finagle/ssl/SslConfigurations.scala]: style nit: please revert lining up of the `=>`s (here and below)

File: finagle-core/src/main/scala/com/twitter/finagle/ssl/TrustCredentials.scala [hunk 6]
```diff
@@ -34,4 +36,11 @@ object TrustCredentials {
    * in PEM format.
    */
   case class CertCollection(file: File) extends TrustCredentials
+
+
+  /**
+   *
+   * @param trustManagerFactory
+   */
+  case class FromTrustManagerFactory(trustManagerFactory: TrustManagerFactory) extends TrustCredentials
```
Comment (kevinoliver) [inline finagle-core/src/main/scala/com/twitter/finagle/ssl/TrustCredentials.scala]: same here for scaladoc and naming.

File: finagle-netty4/src/main/scala/com/twitter/finagle/netty4/ssl/Netty4SslConfigurations.scala [hunk 7]
```diff
@@ -25,6 +25,11 @@ private[finagle] object Netty4SslConfigurations {
    * @note TrustCredentials.Unspecified does not change the builder,
    * as it is not possible to set the trustManager to use the system
    * trust credentials, like with the JDK engine factories.
+   *
+   * @todo Is the note above still true?
+   *       AFAIK, we can do
```
Comment (kevinoliver) [inline finagle-netty4/src/main/scala/com/twitter/finagle/netty4/ssl/Netty4SslConfigurations.scala]: how can we answer this?

File: finagle-netty4/src/test/scala/com/twitter/finagle/netty4/ssl/server/Netty4ServerEngineFactoryTest.scala [hunk 8]
```diff
@@ -112,6 +116,26 @@ class Netty4ServerEngineFactoryTest extends FunSuite {
     }
   }
 
+  test("config with KeyManager succeeds") {
+    // ToDo: Should we also create tests / build the functionality that the keymanagers are rejected if containing expired certs
```
Comment (kevinoliver) [inline finagle-netty4/src/test/scala/com/twitter/finagle/netty4/ssl/server/Netty4ServerEngineFactoryTest.scala]: how hard is this? 

i defer to @ryanoneill

File: finagle-core/src/main/scala/com/twitter/finagle/ssl/KeyCredentials.scala (hunk 4 above)
Comment (spockz) [reply finagle-core/src/main/scala/com/twitter/finagle/ssl/KeyCredentials.scala]: Calling this class `KeyManagerFactory` as well makes it clash with the argument type. I'm ambivalent in overriding names like that especially when making it usable from Java which does not have import renaming.

File: finagle-core/src/main/java/com/twitter/finagle/ssl/KeyCredentialsConfig.java (hunk 1 above)
Comment (spockz) [reply finagle-core/src/main/java/com/twitter/finagle/ssl/KeyCredentialsConfig.java]: Done

File: finagle-netty4/src/test/scala/com/twitter/finagle/netty4/ssl/server/Netty4ServerEngineFactoryTest.scala (hunk 8 above)
Comment (spockz) [reply finagle-netty4/src/test/scala/com/twitter/finagle/netty4/ssl/server/Netty4ServerEngineFactoryTest.scala]: I have not added the test. Depending on how the KeyManagerFactory is constructed the check is already performed there.

File: finagle-core/src/main/scala/com/twitter/finagle/ssl/SslConfigurations.scala [hunk 9]
```diff
@@ -151,6 +155,11 @@ private[ssl] object SslConfigurations {
           "KeyCredentials.CertKeyAndChain",
           engineFactoryName
         )
+      case KeyCredentials.KeyManagerFactory(_) =>
+        throw SslConfigurationException.notSupported(
+          "KeyCredentials.FromKeyManager",
```
Comment (ryanoneill) [inline finagle-core/src/main/scala/com/twitter/finagle/ssl/SslConfigurations.scala]: Please update to the renamed version.

File: finagle-core/src/main/scala/com/twitter/finagle/ssl/TrustCredentials.scala [hunk 10]
```diff
@@ -30,8 +31,19 @@ object TrustCredentials {
    * The collection of certificates which should be used in
    * verifying a remote peer's credentials.
    *
-   * @file A file containing a collection of X.509 certificates
-   * in PEM format.
+   * @param file A file containing a collection of X.509 certificates
```
Comment (ryanoneill) [inline finagle-core/src/main/scala/com/twitter/finagle/ssl/TrustCredentials.scala]: Thanks

File: finagle-netty4/src/main/scala/com/twitter/finagle/netty4/ssl/client/Netty4ClientSslConfigurations.scala [hunk 11]
```diff
@@ -48,7 +48,7 @@ private[finagle] object Netty4ClientSslConfigurations {
     val withKey = keyCredentials match {
       case KeyCredentials.Unspecified =>
         Return(builder) // Do Nothing
-      case KeyCredentials.CertAndKey(certFile, keyFile) =>
+      case KeyCredentials.CertAndKey(certFile, keyFile)                 =>
```
Comment (ryanoneill) [inline finagle-netty4/src/main/scala/com/twitter/finagle/netty4/ssl/client/Netty4ClientSslConfigurations.scala]: Please undo the spacing change here.

File: finagle-netty4/src/main/scala/com/twitter/finagle/netty4/ssl/client/Netty4ClientSslConfigurations.scala [hunk 12]
```diff
@@ -59,6 +59,8 @@ private[finagle] object Netty4ClientSslConfigurations {
           cert <- new X509CertificateFile(certFile).readX509Certificate()
           chain <- new X509CertificateFile(chainFile).readX509Certificates()
         } yield builder.keyManager(key, cert +: chain: _*)
+      case KeyCredentials.KeyManagerFactory(keyManagerFactory)          =>
```
Comment (ryanoneill) [inline finagle-netty4/src/main/scala/com/twitter/finagle/netty4/ssl/client/Netty4ClientSslConfigurations.scala]: Please move the '=>' back near the closing parenthesis.

File: finagle-netty4/src/main/scala/com/twitter/finagle/netty4/ssl/server/Netty4ServerSslConfigurations.scala [hunk 13]
```diff
@@ -54,7 +50,7 @@ private[finagle] object Netty4ServerSslConfigurations {
           "KeyCredentials.Unspecified",
           "Netty4ServerEngineFactory"
         )
-      case KeyCredentials.CertAndKey(certFile, keyFile) =>
+      case KeyCredentials.CertAndKey(certFile, keyFile)                 =>
```
Comment (ryanoneill) [inline finagle-netty4/src/main/scala/com/twitter/finagle/netty4/ssl/server/Netty4ServerSslConfigurations.scala]: Please undo the spacing change.

File: finagle-netty4/src/main/scala/com/twitter/finagle/netty4/ssl/server/Netty4ServerSslConfigurations.scala [hunk 14]
```diff
@@ -65,6 +61,8 @@ private[finagle] object Netty4ServerSslConfigurations {
           cert <- new X509CertificateFile(certFile).readX509Certificate()
           chain <- new X509CertificateFile(chainFile).readX509Certificates()
         } yield SslContextBuilder.forServer(key, cert +: chain: _*)
+      case KeyCredentials.KeyManagerFactory(keyManagerFactory)          =>
```
Comment (ryanoneill) [inline finagle-netty4/src/main/scala/com/twitter/finagle/netty4/ssl/server/Netty4ServerSslConfigurations.scala]: Please move the '=>' back near the closing parenthesis.

File: finagle-netty4/src/test/scala/com/twitter/finagle/netty4/ssl/server/Netty4ServerEngineFactoryTest.scala [hunk 15]
```diff
@@ -112,6 +114,39 @@ class Netty4ServerEngineFactoryTest extends FunSuite {
...
+
+    val config = SslServerConfiguration(trustCredentials = trustCredentials, keyCredentials = keyCredentials)
+    val engine = factory(config)
+    val sslEngine = engine.self
+
+    assert(sslEngine != null)
+  }
+
+  test("config with only KeyCredentials succeeds") {
+    // ToDo: Should we also create tests / build the functionality that the keymanagers are rejected if containing expired certs
```
Comment (ryanoneill) [inline finagle-netty4/src/test/scala/com/twitter/finagle/netty4/ssl/server/Netty4ServerEngineFactoryTest.scala]: Remove. I think it's ok to not do this.

File: finagle-netty4/src/main/scala/com/twitter/finagle/netty4/ssl/client/Netty4ClientSslConfigurations.scala (hunk 12 above)
Comment (ryanoneill) [inline finagle-netty4/src/main/scala/com/twitter/finagle/netty4/ssl/client/Netty4ClientSslConfigurations.scala]: Can you add a note to the comments for this method that indicates that unlike Cert/Key Cert/Key/Chain, that using KeyManagerFactory does not ensure that the certificates contained within the KeyManagerFactory are not expired.

File: finagle-netty4/src/main/scala/com/twitter/finagle/netty4/ssl/server/Netty4ServerSslConfigurations.scala (hunk 14 above)
Comment (ryanoneill) [inline finagle-netty4/src/main/scala/com/twitter/finagle/netty4/ssl/server/Netty4ServerSslConfigurations.scala]: Can you add a note to the comments for this method that indicates that unlike Cert/Key Cert/Key/Chain, that using KeyManagerFactory does not ensure that the certificates contained within the KeyManagerFactory are not expired.

File: finagle-core/src/main/java/com/twitter/finagle/ssl/KeyCredentialsConfig.java [hunk 16]
```diff
@@ -33,4 +34,11 @@ public static KeyCredentials certKeyAndChain(
     return new KeyCredentials.CertKeyAndChain(certificateFile, keyFile, caCertificateFile);
   }
 
+  /**
+   * See {@link KeyCredentials.KeyManagerFactory}
+   * @see KeyCredentials.KeyManagerFactory
```
[the remark this answers is not in the document]
Comment (spockz) [reply finagle-core/src/main/java/com/twitter/finagle/ssl/KeyCredentialsConfig.java]: Okay. Why don't we use `@see`?

[the remark this answers is not in the document]
Comment (ryanoneill) [reply finagle-core/src/main/java/com/twitter/finagle/ssl/KeyCredentialsConfig.java]: TBH, I'm not sure which is better, but I'd prefer it to be consistent for now.

**Comment by yufangong** [comment]
Thanks @spockz , this is merged in https://github.com/twitter/finagle/commit/c863ca0b606455d78350588595b3190165efe26b.

**Pull request closed without merging**